domain/src/main/java/com/thoughtworks/go/domain/materials/git/GitCommand.java

```diff
@@ -56,95 +59,97 @@ public GitCommand(String materialFingerprint, File workingDir, String branch, bo
-        return clone(outputStreamConsumer, url, Integer.MAX_VALUE);
+    public String remoteBranch() {
+        return RefSpecHelper.remoteBranch(branch);
     }
 
-    // Clone repository from url with specified depth.
-    // Special depth 2147483647 (Integer.MAX_VALUE) are treated as full clone
-    public int clone(ConsoleOutputStreamConsumer outputStreamConsumer, String url, Integer depth) {
-        CommandLine gitClone = cloneCommand();
+    public String fullUpstreamRef() {
+        return RefSpecHelper.fullUpstreamRef(branch);
+    }
 
-        if (depth < Integer.MAX_VALUE) {
-            gitClone.withArg(String.format("--depth=%s", depth));
-        }
-        gitClone.withArg(new UrlArgument(url)).withArg(workingDir.getAbsolutePath());
+    public String expandRefSpec() {
+        return RefSpecHelper.expandRefSpec(branch);
+    }
 
-        return run(gitClone, outputStreamConsumer);
+    public boolean hasRefSpec() {
+        return RefSpecHelper.hasRefSpec(branch);
     }
 
-    private CommandLine cloneCommand() {
-        return git()
-                .withArg("clone")
-                .withArg(String.format("--branch=%s", branch));
+    public String getConfigValue(String key) {
+        return runOrBomb(git_C().withArgs("config", "--get-all", key)).outputAsString();
     }
 
-    public List<Modification> latestModification() {
-        return gitLog("-1", "--date=iso", "--no-decorate", "--pretty=medium", "--no-color", remoteBranch());
+    public int cloneWithNoCheckout(ConsoleOutputStreamConsumer outputStreamConsumer, String url) {
+        CommandLine gitClone = cloneCommand().
+                when(!hasRefSpec(), git -> git.withArgs("--branch", branch)).
+                withArg("--no-checkout").
+                withArg(new UrlArgument(url)).
+                withArg(workingDir.getAbsolutePath());
+
+        if (!hasRefSpec()) {
+            return run(gitClone, outputStreamConsumer);
+        }
+
+        final String abbrevBranch = localBranch();
+        final String fullLocalRef = abbrevBranch.startsWith("refs/") ? abbrevBranch : REFS_HEADS + abbrevBranch;
 
+        return runCascade(outputStreamConsumer,
+                gitClone,
+                git_C().withArgs("config", "--replace-all", "remote.origin.fetch", "+" + expandRefSpec()),
+                git_C().withArgs("fetch", "--prune", "--recurse-submodules=no"),
+                // Important to create a "real" local branch and not just use `symbolic-ref`
+                // to update HEAD in order to ensure that GitMaterial#isBranchEqual() passes;
+                // failing this check will cause the working directory to be obliterated and we
+                // will re-clone the given repository every time. Yikes!
+                git_C().withArgs("branch", "-f", abbrevBranch, remoteBranch()),
+                git_C().withArgs("symbolic-ref", "HEAD", fullLocalRef)
+        );
```

Comment:
This and `clone()` are the main changes here (other parts are IDEA cleanup). Cloning to accommodate for refspecs is a multi-step process involving manually configuring fetch behavior and explicit branch creation and switching.